src/shared/Microsoft.Git.CredentialManager.Tests/SettingsTests.cs

```diff
@@ -191,6 +191,266 @@ public void Settings_IsSecretTracingEnabled_EnvarFalsey_ReturnsFalse()
+            var remoteUri = new Uri(remoteUrl);
+
+            var expectedValue = new Uri("http://john.doe:letmein123@proxy.example.com");
+
+            var envars = new EnvironmentVariables(new Dictionary<string, string>
+            {
+                [Constants.EnvironmentVariables.GcmHttpProxy] = expectedValue.ToString()
+            });
+            var git = new TestGit();
+
+            var settings = new Settings(envars, git)
+            {
+                RemoteUri = remoteUri,
+                RepositoryPath = repositoryPath
+            };
+            Uri actualValue = settings.GetProxyConfiguration(out bool actualIsDeprecated);
+
+            Assert.Equal(expectedValue, actualValue);
+            Assert.True(actualIsDeprecated);
+        }
+
+        [Fact]
+        public void Settings_ProxyConfiguration_Precedence_ReturnsValue()
+        {
+            // 1. GCM proxy Git configuration (deprecated)
+            //      credential.httpsProxy
+            //      credential.httpProxy
+            // 2. Standard Git configuration
+            //      http.proxy
+            // 3. cURL environment variables
+            //      HTTPS_PROXY
+            //      HTTP_PROXY
+            //      ALL_PROXY
+            // 4. GCM proxy environment variable (deprecated)
+            //      GCM_HTTP_PROXY
+
+            const string repositoryPath = "/tmp/repos/foo/.git";
+            const string remoteUrl = "http://example.com/foo.git";
+            var remoteUri = new Uri(remoteUrl);
+
+            var value1 = new Uri("http://proxy1.example.com");
+            var value2 = new Uri("http://proxy2.example.com");
+            var value3 = new Uri("http://proxy3.example.com");
+            var value4 = new Uri("http://proxy4.example.com");
+
+            var envarDict = new Dictionary<string, string>();
+            var configDict = new Dictionary<string, string>();
+
+            void RunTest(Uri expectedValue)
+            {
+                var settings = new Settings(new EnvironmentVariables(envarDict), new TestGit(configDict))
+                {
+                    RemoteUri = remoteUri,
+                    RepositoryPath = repositoryPath
+                };
+                Uri actualValue = settings.GetProxyConfiguration(out bool actualIsDeprecated);
+                Assert.Equal(expectedValue, actualValue);
+            }
+
+             // Test case 1: cURL environment variables > GCM_HTTP_PROXY
+            envarDict[Constants.EnvironmentVariables.GcmHttpProxy] = value1.ToString();
+            envarDict[Constants.EnvironmentVariables.CurlHttpProxy] = value2.ToString();
+            RunTest(value2);
+
+             // Test case 2: http.proxy > cURL environment variables
+            configDict[$"{Constants.GitConfiguration.Http.SectionName}.{Constants.GitConfiguration.Http.Proxy}"] = value3.ToString();
+            RunTest(value3);
+
+             // Test case 3: credential.httpProxy > http.proxy
```

Comment:
did we skip 4 cause it's being depricated?  If so I'm fine with that.

Reply:
We did not skip any. This test is checking the precedence of the 4 options:

```
                     1            2            3
credential.httpProxy > HTTP_PROXY > http.proxy > GCM_HTTP_PROXY
```

There are three inequalities to check for these four options.